File: LeetCode/21.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

def mergeTwoLists(l1, l2):
    """
    :type l1: ListNode
    :type l2: ListNode
    :rtype: ListNode
    """
    if not l1: return l2
    if not l2: return l1

    if l1.val <= l2.val:
    	l1.next = mergeTwoLists(l1.next, l2)
    	return l1
    else:
    	l2.next = mergeTwoLists(l1, l2.next)
    	return l2

File: LeetCode/21.java
/*
21. Merge Two Sorted Lists

Merge two sorted linked lists and return it as a new list. 
The new list should be made by splicing together the nodes of the first two lists.
*/
/*
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        // corner cases first
        // null lists in this case
        if(l1 == null) return l2;
        if(l2 == null) return l1;
        
        // Use recursion
        // Every time you find the smaller one as the merge head meanwhile it's the last recursive procession's next node. 
        ListNode newHead;
        if(l1.val <= l2.val) {
            newHead = l1;
            newHead.next = mergeTwoLists(l1.next, l2);
        } else {
            newHead = l2;
            newHead.next = mergeTwoLists(l2.next, l1);
        }
        return newHead;
    }
    
}

